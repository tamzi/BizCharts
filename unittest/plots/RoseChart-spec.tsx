import React from 'react';
import RoseChart from '../../src/plots/RoseChart';
import { render, cleanup } from '@testing-library/react';

const MOCK_DATA = [{
  "Date": "2010-01",
  "scales": 1998
},
{
  "Date": "2010-02",
  "scales": 1250
},
{
  "Date": "2010-03",
  "scales": 1720
}];

const basicCfg = {
  padding: [90, 90, 90, 90],
  data: MOCK_DATA,
  xField: "Date",
  yField: "scales",
  seriesField: "Date",
  autoFit: true,
  width: 600,
  height: 500,
};

describe('Plots-RoseChart', () => {
  test('新版本玫瑰图xField*yField*seriesFieldte', () => {
    let chart = null;
    render(<RoseChart
      {...basicCfg}
      title="新版本玫瑰图xField*yField*seriesField"
      onGetG2Instance={
        (c) => chart = c
      }
    />)
    expect(chart.options).toMatchSnapshot();
    cleanup();
  });

  test('老版本玫瑰图categoryField*radiusField*Date', () => {
    let chart = null;
    // @ts-ignore
    render(<RoseChart
      padding={[90, 90, 90, 90]}
      data={MOCK_DATA}
      categoryField="Date"
      radiusField="scales"
      colorField="Date"
      autoFit
      width={600}
      height={500}
      title="老版本categoryField*radiusField*Date"
      onGetG2Instance={
        (c) => chart = c
      }
    />)
    expect(chart.options).toMatchSnapshot();
    cleanup();
  });

  test('基础玫瑰图', () => {
    let chart = null;
    render(<RoseChart
      {...basicCfg}
      title="基础玫瑰图"
      pixelRatio={3}
      renderer="svg"
      meta={{
        date: {
          formatter: val => `${val}%`
        }
      }}
      radius={0.5}
      innerRadius={0.1}
      // color={['red', 'green', 'yellow']}
      color={({ Date }) => {
        if (Date === '2010-02') return 'red';
        return 'blue';
      }}
      sectorStyle={{
        stroke: 'black',
        lineDash: [5, 5],
      }}
      onGetG2Instance={
        (c) => chart = c
      }
    />);
    expect(chart.options).toMatchSnapshot();
    cleanup();
  });


  test('legend-玫瑰图', () => {
    let chart = null;
    const fn = jest.fn(val => `${val}%`);
    const fn2 = jest.fn(val => `${val}%`);
    render(<RoseChart
      title={{
        visible: true,
        text: "legend-玫瑰图",
      }}
      {...basicCfg}
      legend={{
        position: "bottom",
        formatter: fn, //  text 的优先度大于formatter
        offsetX: 40,
        offsetY: 2,
        title: {
          text: 'legend title',
          style: {
            fill: "red",
          },
        },
        marker: {
          symbol: "circle",
        },
        text: {
          style: {
            fill: "pink",
            stroke: "pink",
          },
          formatter: fn2,
        },
      }}
      onGetG2Instance={
        (c) => chart = c
      }
    />);
    expect(fn).not.toHaveBeenCalled();
    expect(fn2).toHaveBeenCalled();
    expect(chart.options).toMatchSnapshot();
    cleanup();
  });


  test('tooltip-玫瑰图', () => {
    let chart = null;
    render(<RoseChart
      title="tooltip-玫瑰图"
      {...basicCfg}
      tooltip={{
        visible: true,
        offset: 10,
        shared: true,
        showCrosshairs: true,
        crosshairs: { type: 'y' },
        fields: ['scales'],
      }}
      onGetG2Instance={
        (c) => chart = c
      }
    />);
    expect(chart.options).toMatchSnapshot();
    cleanup();
  });

  test('label-玫瑰图', () => {
    let chart = null;
    render(<RoseChart
      title="label-玫瑰图"
      {...basicCfg}
      label={{
        visible: true,
        autoRotate: true,
        formatter: ({ Date }) => `${Date}val`,
        offsetX: 6,
        offsetY: 6,
        style: {
          fill: 'red',
        }
      }}
      onGetG2Instance={
        (c) => chart = c
      }
    />);
    expect(chart.options).toMatchSnapshot();
    cleanup();
  });

  test('label-inner-bug', () => {
    let chart = null;
    render(<RoseChart
      title="label-玫瑰图"
      {...basicCfg}
      label={{
        visible: true,
        type: 'inner',
      }}
      onGetG2Instance={
        (c) => chart = c
      }
    />);
    expect(chart.options).toMatchSnapshot();
    cleanup();
  });

  test('label-outer-bug', () => {
    let chart = null;
    render(<RoseChart
      title="label-玫瑰图"
      {...basicCfg}
      label={{
        visible: true,
        type: 'outer',
      }}
      onGetG2Instance={
        (c) => chart = c
      }
    />);
    expect(chart.options).toMatchSnapshot();
    cleanup();
  });

});
