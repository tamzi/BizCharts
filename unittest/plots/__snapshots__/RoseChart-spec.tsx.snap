// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Plots-RoseChart label-inner-bug 1`] = `
Object {
  "animation": true,
  "appendPadding": Array [
    10,
    5,
    10,
    10,
  ],
  "autoFit": true,
  "data": Array [
    Object {
      "Date": "2010-01",
      "scales": 1998,
    },
    Object {
      "Date": "2010-02",
      "scales": 1250,
    },
    Object {
      "Date": "2010-03",
      "scales": 1720,
    },
  ],
  "height": 500,
  "interactions": Array [
    Object {
      "type": "active-region",
    },
  ],
  "label": Object {
    "layout": Object {
      "type": "limit-in-shape",
    },
    "offset": -15,
    "visible": true,
  },
  "legend": Object {
    "offsetX": -10,
    "position": "right",
  },
  "padding": Array [
    90,
    90,
    90,
    90,
  ],
  "renderer": "canvas",
  "sectorStyle": Object {
    "lineWidth": 1,
    "stroke": "#fff",
  },
  "seriesField": "Date",
  "tooltip": Object {
    "shared": true,
    "showMarkers": false,
  },
  "width": 600,
  "xAxis": false,
  "xField": "Date",
  "yAxis": false,
  "yField": "scales",
}
`;

exports[`Plots-RoseChart label-outer-bug 1`] = `
Object {
  "animation": true,
  "appendPadding": Array [
    10,
    5,
    10,
    10,
  ],
  "autoFit": true,
  "data": Array [
    Object {
      "Date": "2010-01",
      "scales": 1998,
    },
    Object {
      "Date": "2010-02",
      "scales": 1250,
    },
    Object {
      "Date": "2010-03",
      "scales": 1720,
    },
  ],
  "height": 500,
  "interactions": Array [
    Object {
      "type": "active-region",
    },
  ],
  "label": Object {
    "layout": Object {
      "type": "limit-in-shape",
    },
    "visible": true,
  },
  "legend": Object {
    "offsetX": -10,
    "position": "right",
  },
  "padding": Array [
    90,
    90,
    90,
    90,
  ],
  "renderer": "canvas",
  "sectorStyle": Object {
    "lineWidth": 1,
    "stroke": "#fff",
  },
  "seriesField": "Date",
  "tooltip": Object {
    "shared": true,
    "showMarkers": false,
  },
  "width": 600,
  "xAxis": false,
  "xField": "Date",
  "yAxis": false,
  "yField": "scales",
}
`;

exports[`Plots-RoseChart label-玫瑰图 1`] = `
Object {
  "animation": true,
  "appendPadding": Array [
    10,
    5,
    10,
    10,
  ],
  "autoFit": true,
  "data": Array [
    Object {
      "Date": "2010-01",
      "scales": 1998,
    },
    Object {
      "Date": "2010-02",
      "scales": 1250,
    },
    Object {
      "Date": "2010-03",
      "scales": 1720,
    },
  ],
  "height": 500,
  "interactions": Array [
    Object {
      "type": "active-region",
    },
  ],
  "label": Object {
    "autoRotate": true,
    "formatter": [Function],
    "layout": Object {
      "type": "limit-in-shape",
    },
    "offsetX": 6,
    "offsetY": 6,
    "style": Object {
      "fill": "red",
    },
    "visible": true,
  },
  "legend": Object {
    "offsetX": -10,
    "position": "right",
  },
  "padding": Array [
    90,
    90,
    90,
    90,
  ],
  "renderer": "canvas",
  "sectorStyle": Object {
    "lineWidth": 1,
    "stroke": "#fff",
  },
  "seriesField": "Date",
  "tooltip": Object {
    "shared": true,
    "showMarkers": false,
  },
  "width": 600,
  "xAxis": false,
  "xField": "Date",
  "yAxis": false,
  "yField": "scales",
}
`;

exports[`Plots-RoseChart legend-玫瑰图 1`] = `
Object {
  "animation": true,
  "appendPadding": Array [
    10,
    5,
    10,
    10,
  ],
  "autoFit": true,
  "data": Array [
    Object {
      "Date": "2010-01",
      "scales": 1998,
    },
    Object {
      "Date": "2010-02",
      "scales": 1250,
    },
    Object {
      "Date": "2010-03",
      "scales": 1720,
    },
  ],
  "height": 500,
  "interactions": Array [
    Object {
      "type": "active-region",
    },
  ],
  "label": Object {
    "layout": Object {
      "type": "limit-in-shape",
    },
  },
  "legend": Object {
    "formatter": [MockFunction],
    "itemName": Object {
      "formatter": [MockFunction] {
        "calls": Array [
          Array [
            "2010-01",
            Object {
              "id": "2010-01",
              "marker": Object {
                "spacing": 8,
                "style": Object {
                  "fill": "#5B8FF9",
                  "r": 4.5,
                },
                "symbol": "circle",
              },
              "name": "2010-01",
              "unchecked": false,
              "value": "2010-01",
            },
            0,
          ],
          Array [
            "2010-02",
            Object {
              "id": "2010-02",
              "marker": Object {
                "spacing": 8,
                "style": Object {
                  "fill": "#5AD8A6",
                  "r": 4.5,
                },
                "symbol": "circle",
              },
              "name": "2010-02",
              "unchecked": false,
              "value": "2010-02",
            },
            1,
          ],
          Array [
            "2010-03",
            Object {
              "id": "2010-03",
              "marker": Object {
                "spacing": 8,
                "style": Object {
                  "fill": "#5D7092",
                  "r": 4.5,
                },
                "symbol": "circle",
              },
              "name": "2010-03",
              "unchecked": false,
              "value": "2010-03",
            },
            2,
          ],
          Array [
            "2010-01",
            Object {
              "id": "2010-01",
              "marker": Object {
                "spacing": 8,
                "style": Object {
                  "fill": "#5B8FF9",
                  "r": 4.5,
                },
                "symbol": "circle",
              },
              "name": "2010-01",
              "unchecked": false,
              "value": "2010-01",
            },
            0,
          ],
          Array [
            "2010-02",
            Object {
              "id": "2010-02",
              "marker": Object {
                "spacing": 8,
                "style": Object {
                  "fill": "#5AD8A6",
                  "r": 4.5,
                },
                "symbol": "circle",
              },
              "name": "2010-02",
              "unchecked": false,
              "value": "2010-02",
            },
            1,
          ],
          Array [
            "2010-03",
            Object {
              "id": "2010-03",
              "marker": Object {
                "spacing": 8,
                "style": Object {
                  "fill": "#5D7092",
                  "r": 4.5,
                },
                "symbol": "circle",
              },
              "name": "2010-03",
              "unchecked": false,
              "value": "2010-03",
            },
            2,
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": "2010-01%",
          },
          Object {
            "type": "return",
            "value": "2010-02%",
          },
          Object {
            "type": "return",
            "value": "2010-03%",
          },
          Object {
            "type": "return",
            "value": "2010-01%",
          },
          Object {
            "type": "return",
            "value": "2010-02%",
          },
          Object {
            "type": "return",
            "value": "2010-03%",
          },
        ],
      },
      "style": Object {
        "fill": "pink",
        "stroke": "pink",
      },
    },
    "marker": Object {
      "symbol": "circle",
    },
    "offsetX": 40,
    "offsetY": 2,
    "position": "bottom",
    "text": Object {
      "formatter": [MockFunction] {
        "calls": Array [
          Array [
            "2010-01",
            Object {
              "id": "2010-01",
              "marker": Object {
                "spacing": 8,
                "style": Object {
                  "fill": "#5B8FF9",
                  "r": 4.5,
                },
                "symbol": "circle",
              },
              "name": "2010-01",
              "unchecked": false,
              "value": "2010-01",
            },
            0,
          ],
          Array [
            "2010-02",
            Object {
              "id": "2010-02",
              "marker": Object {
                "spacing": 8,
                "style": Object {
                  "fill": "#5AD8A6",
                  "r": 4.5,
                },
                "symbol": "circle",
              },
              "name": "2010-02",
              "unchecked": false,
              "value": "2010-02",
            },
            1,
          ],
          Array [
            "2010-03",
            Object {
              "id": "2010-03",
              "marker": Object {
                "spacing": 8,
                "style": Object {
                  "fill": "#5D7092",
                  "r": 4.5,
                },
                "symbol": "circle",
              },
              "name": "2010-03",
              "unchecked": false,
              "value": "2010-03",
            },
            2,
          ],
          Array [
            "2010-01",
            Object {
              "id": "2010-01",
              "marker": Object {
                "spacing": 8,
                "style": Object {
                  "fill": "#5B8FF9",
                  "r": 4.5,
                },
                "symbol": "circle",
              },
              "name": "2010-01",
              "unchecked": false,
              "value": "2010-01",
            },
            0,
          ],
          Array [
            "2010-02",
            Object {
              "id": "2010-02",
              "marker": Object {
                "spacing": 8,
                "style": Object {
                  "fill": "#5AD8A6",
                  "r": 4.5,
                },
                "symbol": "circle",
              },
              "name": "2010-02",
              "unchecked": false,
              "value": "2010-02",
            },
            1,
          ],
          Array [
            "2010-03",
            Object {
              "id": "2010-03",
              "marker": Object {
                "spacing": 8,
                "style": Object {
                  "fill": "#5D7092",
                  "r": 4.5,
                },
                "symbol": "circle",
              },
              "name": "2010-03",
              "unchecked": false,
              "value": "2010-03",
            },
            2,
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": "2010-01%",
          },
          Object {
            "type": "return",
            "value": "2010-02%",
          },
          Object {
            "type": "return",
            "value": "2010-03%",
          },
          Object {
            "type": "return",
            "value": "2010-01%",
          },
          Object {
            "type": "return",
            "value": "2010-02%",
          },
          Object {
            "type": "return",
            "value": "2010-03%",
          },
        ],
      },
      "style": Object {
        "fill": "pink",
        "stroke": "pink",
      },
    },
    "title": Object {
      "style": Object {
        "fill": "red",
      },
      "text": "legend title",
    },
  },
  "padding": Array [
    90,
    90,
    90,
    90,
  ],
  "renderer": "canvas",
  "sectorStyle": Object {
    "lineWidth": 1,
    "stroke": "#fff",
  },
  "seriesField": "Date",
  "tooltip": Object {
    "shared": true,
    "showMarkers": false,
  },
  "width": 600,
  "xAxis": false,
  "xField": "Date",
  "yAxis": false,
  "yField": "scales",
}
`;

exports[`Plots-RoseChart tooltip-玫瑰图 1`] = `
Object {
  "animation": true,
  "appendPadding": Array [
    10,
    5,
    10,
    10,
  ],
  "autoFit": true,
  "data": Array [
    Object {
      "Date": "2010-01",
      "scales": 1998,
    },
    Object {
      "Date": "2010-02",
      "scales": 1250,
    },
    Object {
      "Date": "2010-03",
      "scales": 1720,
    },
  ],
  "height": 500,
  "interactions": Array [
    Object {
      "type": "active-region",
    },
  ],
  "label": Object {
    "layout": Object {
      "type": "limit-in-shape",
    },
  },
  "legend": Object {
    "offsetX": -10,
    "position": "right",
  },
  "padding": Array [
    90,
    90,
    90,
    90,
  ],
  "renderer": "canvas",
  "sectorStyle": Object {
    "lineWidth": 1,
    "stroke": "#fff",
  },
  "seriesField": "Date",
  "tooltip": Object {
    "crosshairs": Object {
      "type": "y",
    },
    "fields": Array [
      "scales",
    ],
    "offset": 10,
    "shared": true,
    "showCrosshairs": true,
    "showMarkers": false,
    "visible": true,
  },
  "width": 600,
  "xAxis": false,
  "xField": "Date",
  "yAxis": false,
  "yField": "scales",
}
`;

exports[`Plots-RoseChart 基础玫瑰图 1`] = `
Object {
  "animation": true,
  "appendPadding": Array [
    10,
    5,
    10,
    10,
  ],
  "autoFit": true,
  "color": [Function],
  "data": Array [
    Object {
      "Date": "2010-01",
      "scales": 1998,
    },
    Object {
      "Date": "2010-02",
      "scales": 1250,
    },
    Object {
      "Date": "2010-03",
      "scales": 1720,
    },
  ],
  "height": 500,
  "innerRadius": 0.1,
  "interactions": Array [
    Object {
      "type": "active-region",
    },
  ],
  "label": Object {
    "layout": Object {
      "type": "limit-in-shape",
    },
  },
  "legend": Object {
    "offsetX": -10,
    "position": "right",
  },
  "meta": Object {
    "date": Object {
      "formatter": [Function],
    },
  },
  "padding": Array [
    90,
    90,
    90,
    90,
  ],
  "pixelRatio": 3,
  "radius": 0.5,
  "renderer": "svg",
  "sectorStyle": Object {
    "lineDash": Array [
      5,
      5,
    ],
    "lineWidth": 1,
    "stroke": "black",
  },
  "seriesField": "Date",
  "tooltip": Object {
    "shared": true,
    "showMarkers": false,
  },
  "width": 600,
  "xAxis": false,
  "xField": "Date",
  "yAxis": false,
  "yField": "scales",
}
`;

exports[`Plots-RoseChart 新版本玫瑰图xField*yField*seriesFieldte 1`] = `
Object {
  "animation": true,
  "appendPadding": Array [
    10,
    5,
    10,
    10,
  ],
  "autoFit": true,
  "data": Array [
    Object {
      "Date": "2010-01",
      "scales": 1998,
    },
    Object {
      "Date": "2010-02",
      "scales": 1250,
    },
    Object {
      "Date": "2010-03",
      "scales": 1720,
    },
  ],
  "height": 500,
  "interactions": Array [
    Object {
      "type": "active-region",
    },
  ],
  "label": Object {
    "layout": Object {
      "type": "limit-in-shape",
    },
  },
  "legend": Object {
    "offsetX": -10,
    "position": "right",
  },
  "padding": Array [
    90,
    90,
    90,
    90,
  ],
  "renderer": "canvas",
  "sectorStyle": Object {
    "lineWidth": 1,
    "stroke": "#fff",
  },
  "seriesField": "Date",
  "tooltip": Object {
    "shared": true,
    "showMarkers": false,
  },
  "width": 600,
  "xAxis": false,
  "xField": "Date",
  "yAxis": false,
  "yField": "scales",
}
`;

exports[`Plots-RoseChart 老版本玫瑰图categoryField*radiusField*Date 1`] = `
Object {
  "animation": true,
  "appendPadding": Array [
    10,
    5,
    10,
    10,
  ],
  "autoFit": true,
  "categoryField": "Date",
  "colorField": "Date",
  "data": Array [
    Object {
      "Date": "2010-01",
      "scales": 1998,
    },
    Object {
      "Date": "2010-02",
      "scales": 1250,
    },
    Object {
      "Date": "2010-03",
      "scales": 1720,
    },
  ],
  "height": 500,
  "interactions": Array [
    Object {
      "type": "active-region",
    },
  ],
  "label": Object {
    "layout": Object {
      "type": "limit-in-shape",
    },
  },
  "legend": Object {
    "offsetX": -10,
    "position": "right",
  },
  "padding": Array [
    90,
    90,
    90,
    90,
  ],
  "radiusField": "scales",
  "renderer": "canvas",
  "sectorStyle": Object {
    "lineWidth": 1,
    "stroke": "#fff",
  },
  "seriesField": "Date",
  "tooltip": Object {
    "shared": true,
    "showMarkers": false,
  },
  "width": 600,
  "xAxis": false,
  "xField": "Date",
  "yAxis": false,
  "yField": "scales",
}
`;
